File: app/src/main/java/openfoodfacts/github/scrachx/openfood/category/CategoryRepository.java
package openfoodfacts.github.scrachx.openfood.category;

import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import openfoodfacts.github.scrachx.openfood.category.mapper.CategoryMapper;
import openfoodfacts.github.scrachx.openfood.category.model.Category;
import openfoodfacts.github.scrachx.openfood.category.network.CategoryNetworkService;

public class CategoryRepository {
    private final CategoryNetworkService networkService;
    private final CategoryMapper mapper;
    private final AtomicReference<List<Category>> memoryCache;

    public CategoryRepository(CategoryNetworkService networkService, CategoryMapper mapper) {
        this.networkService = networkService;
        this.mapper = mapper;
        memoryCache = new AtomicReference<>();
    }

    public Single<List<Category>> retrieveAll() {
        if (memoryCache.get() != null) {
            return Single.just(memoryCache.get());
        }
        return networkService.getCategories()
                .map(categoryResponse -> mapper.fromNetwork(categoryResponse.getTags()))
                .doOnSuccess(memoryCache::set)
                .subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/category/CategoryRepository.kt
package openfoodfacts.github.scrachx.openfood.category

import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.withContext
import openfoodfacts.github.scrachx.openfood.category.mapper.CategoryMapper
import openfoodfacts.github.scrachx.openfood.category.model.Category
import openfoodfacts.github.scrachx.openfood.category.network.CategoryNetworkService
import java.util.concurrent.atomic.AtomicReference
import javax.inject.Inject
import javax.inject.Singleton

/**
 * This class receives list of all categories using [CategoryNetworkService]
 */
@Singleton
class CategoryRepository @Inject constructor(
    private val networkService: CategoryNetworkService,
    private val mapper: CategoryMapper,
) {
    private val memoryCache = AtomicReference<List<Category>?>()

    /**
     * Calling this function retrieves list of all categories from NetworkService
     */
    suspend fun retrieveAll() = withContext(IO) {
        if (memoryCache.get() != null) {
            memoryCache.get()
        } else {
            networkService.getCategories()
                .let { mapper.fromNetwork(it.tags) }
                .also { memoryCache.set(it) }
        }
    }
}
